### PR TITLE
Updated the socket_opts for aws elasticache

### PR DESCRIPTION
**Pull request by talhaazeemmughal**
- added the `customize_hostname_check` for `socket_opts` before it actually starts the redix instance.

this is related to Issue no #78

### CODE REVIEW

**Comment by ruslandoga** [comment]
👋 @talhaazeemmughal 

If it's not already being done in Redix, there is probably a good reason for it. If not, then I think it should be a PR into Redix instead.

**Comment by talhaazeemmughal** [comment]
> 👋 @talhaazeemmughal
> 
> If it's not already being done in Redix, there is probably a good reason for it. If not, then I think it should be a PR into Redix instead. I think Hammer should accept and pass the options as is into Redix to avoid confusion.

i am just testing things out to get to the the root cause. :D

**Comment by talhaazeemmughal** [comment]
> 👋 @talhaazeemmughal
> 
> If it's not already being done in Redix, there is probably a good reason for it. If not, then I think it should be a PR into Redix instead. I think Hammer should accept and pass the options as is into Redix to avoid confusion.

@ruslandoga ignore the MR for now. If it is something that needs to be passed to the redix then I will move it to the redix repo.

File: lib/hammer_backend_redis.ex [hunk 1]
```diff
@@ -110,6 +110,28 @@ defmodule Hammer.Backend.Redis do
 
     redis_url = Keyword.get(args, :redis_url, nil)
 
+    is_elasticache = Keyword.get(args, :is_elasticache, false)
+    ssl = Keyword.get(redix_config, :ssl, false)
+
+    IO.inspect(ssl, label: "SSL is here => ")
```
Comment (epinault) [inline lib/hammer_backend_redis.ex:116]: please remove IO.inspect

File: lib/hammer_backend_redis.ex [hunk 2]
```diff
@@ -110,6 +110,28 @@ defmodule Hammer.Backend.Redis do
...
+    IO.inspect(is_elasticache, label: "is_elasticache is here => ")
+    redix_config =
+      if ssl and is_elasticache do
+        IO.inspect("inside the if condition")
```
Comment (epinault) [inline lib/hammer_backend_redis.ex:120]: same

File: lib/hammer_backend_redis.ex [hunk 3]
```diff
@@ -110,6 +110,28 @@ defmodule Hammer.Backend.Redis do
...
+          ]
+        ]
+
+        redix_config
+        |> Keyword.put(:socket_opts, socket_opts)
+      else
+        redix_config
+      end
+
+    IO.inspect(redix_config, label: "final redix_config is here =>")
```
Comment (epinault) [inline lib/hammer_backend_redis.ex:133]: same

**Comment by ruslandoga** [comment]
I don't think this needs to be merged. It can be a documentation note instead.

File: lib/hammer_backend_redis.ex (hunk 1 above)
Comment (talhaazeemmughal) [reply lib/hammer_backend_redis.ex]: i did remove them. they were there for investigation purposes.

File: lib/hammer_backend_redis.ex [hunk 4]
```diff
@@ -110,6 +110,28 @@ defmodule Hammer.Backend.Redis do
 
     redis_url = Keyword.get(args, :redis_url, nil)
 
+    is_elasticache = Keyword.get(args, :is_elasticache, false)
+    ssl = Keyword.get(redix_config, :ssl, false)
+
+    IO.inspect(ssl, label: "SSL is here => ")
+    IO.inspect(is_elasticache, label: "is_elasticache is here => ")
```
[the remark this answers is not in the document]
Comment (talhaazeemmughal) [reply lib/hammer_backend_redis.ex]: done

File: lib/hammer_backend_redis.ex (hunk 2 above)
Comment (talhaazeemmughal) [reply lib/hammer_backend_redis.ex]: done

File: lib/hammer_backend_redis.ex (hunk 3 above)
Comment (talhaazeemmughal) [reply lib/hammer_backend_redis.ex]: done

**Comment by talhaazeemmughal** [comment]
@ruslandoga i am not sure, I was passing the `socket_opts` as well but it didn't seemed to work for me. Although this fix does.

**Comment by talhaazeemmughal** [comment]
"""
Hammer.check_rate("test", 60_000, 5)
{:error, %Redix.ConnectionError{reason: :closed}}
"""
the above was the result before even if i was passing `socket_opts` to the hammer_backend_redis.

**Comment by epinault** [comment]
I believe I mentioned to you I was doing a pass through, and you ad configuration issue in your application. I am going to close this for now. A new Major is coming anyway and since it seems to work for you for now I I don t see the need of relleasing it

**Comment by talhaazeemmughal** [comment]
> I believe I mentioned to you I was doing a pass through, and you ad configuration issue in your application. I am going to close this for now. A new Major is coming anyway and since it seems to work for you for now I I don t see the need of relleasing it

@epinault It started working after doing the change i did in this MR.

**Comment by epinault** [comment]
@ruslandoga and I have been rearchitecting this package.  I took over  the project as it was not getting maintenance and needing some work. We are close to release a much better version and I would prefer to focus on that as I have other things I want to add to it and focus my energy on that at so that will solve you issue and  a bunch of other ones too

**Comment by ruslandoga** [comment]
This PR is equivalent to passing

```elixir
Hammer.Backend.Redis.start_link(
  expiry_ms: # ...
  redix_config: [
    # ...
    ssl: true,
    socket_opts: [
      customize_hostname_check: [
        match_fun: :public_key.pkix_verify_hostname_match_fun(:https)
      ]
    ]
  ]
)
```

**Pull request closed without merging**